docs/IR.md

```diff
@@ -82,3 +81,32 @@ The following data types are supported by ONNX. Additional data types can be sup
 |Other|__bool__|Boolean value represent data with only two values, typically _true_ and _false_.|
 |Other|__handle__|Handles are opaque types holding a 64-bit integer.|
 |Collections|__sparse and dense tensor__|Tensors are a generalization of vectors and matrices; whereas vectors have one dimension, and matrices two, tensors can have any number of dimensions, including zero. A zero-dimensional tensor is equivalent to a scalar.|
+
+Extensibility
+-------------
+
+ONNX is expected to evolve over time and provides features that allow users to experiment and implement additions before they are folded into the specifications.
+
+### Metadata
+
+A model allows named metadata strings to be added via its `metadata_props` field, typically for use by tools such as converters, indexers or documentation generators. Names are not prescribed, but some name recommendations are provided for implementations that want to record such concepts.
+
+- `author`: name of the person(s) who authored the model
+- `company`: name of the company or organization that authored the model
+- `converted_from`: if converted from a different format, the name of the source format or framework
+- `license`: a human-readable name for a license if applicable
+- `license_url`: the URL where the license text is provided
+
+### Operators
+
+Operators may be extended via custom domain names in the `opset_import` field.
+
+### Protobuf Fields
```

Comment:
It's not a common practice to have reserved ranges of protobuf fields. I'm not really sure about "reserved for private use" - do we expect people to write their own extensions of ONNX format? If so - what's the way of doing so?

I'm not opposing to this formulation, I'm just really not sure whether and how it'd be used.

Reply:
Really, this is to help people that want to try extending ONNX for some reason to distinguish between "I need a field for my own purposes" (private use, say some "your department gets charged `n` credits per evaluation on company servers") vs. "I'm trying out something that I'd like to share with others" (experimentation) (say, some sort of proposal for handling differential updates). Implements can get traction on this without reaching out to the community, and minimizing their chances of colliding with someone else's extensions.

I don't think it's going to get much usage, but if it ever does, it helps to have this guidance in place.

Reply:
I can understand the private use. But I don't see a need for experimentation range. When you are trying something out, it's private use and can use field #s in the private range. When you are ready to contribute it back, you would submit it as a PR with new field #s in the production range.

Reply:
My concern is that in order to extend protobufs one would need to fork the spec. Protobuf has builtin extension mechanism but unfortunately it's different between proto2 and proto3.

Reply:
@prasanthpul the experimentation range can be used to make sure you can at some point show your work to someone else without immediately colliding with someone's private usage.

@dzhulgakov do you have a different proposal? I imagined that a private extension would indeed fork the spec, if nothing else because it requires a different implementation to actually do something with the new data (and that thing wouldn't just rev transparently with new ONNX updates), so I think it's expected for the scenario not a bug.

An alternative is to kill the protobuf field extensions verbiage to further discourage that. Given how easy it is in practice to drop some new fields in the .proto distribution and get an implementation that is backward compatible with what exists out there, I suspect people will definitely play around with that.

I don't have much personal experience using .proto in production, so I'm happy to hear how this tends to play out and adjust accordingly.

Reply:
@marcelolr if you use experimentation range and share it with someone else who is also using experimentation range, you will have a collision since we are not assigning different experimentation ranges to different parties.

Reply:
@prasanthpul in the experimentation range, you would expect active development so one or another can be updated to coexist. It's also easier to discover because you're sharing your development with others, as opposed to the private range.

People extending a system in a distributed manner are at risk of creating conflicting extensions. Other than discouraging extensions in the first place, I don't see how avoiding any guidance will lead to better outcomes, unless you're arguing for a more sophisticated approach?

Reply:
Discussed with @prasanthpul. We'll just comment on private use range and set up by example the practice of contributing to the community for extensions/experiments.

Reply:
The only other proposal I can think of is to just add `extra_data` string and let people serialize what they want there. In some case by even talking of field ranges we're encouraging extensions. But I don't feel strongly about it and the current solution is fine.

Reply:
let's leave it out for now. its too early for extension